pkg/executor/build.go

```diff
@@ -201,7 +207,7 @@ func (s *stageBuilder) optimize(compositeKey CompositeCache, cfg v1.Config) erro
 			return errors.Wrap(err, "failed to get files used from context")
 		}
 
-		compositeKey, err = s.populateCompositeKey(command, files, compositeKey)
+		compositeKey, err = s.populateCompositeKey(command, files, compositeKey, s.args, cfg.Env)
```

Comment:
It seems like adding `s.args` to the cache key would result in the same problem we have right now where all of the args are in the cache key all the time.

Reply:
I'm confused about the difference between `StageBuilider.args` and `StageBuilder.opts.args`

Reply:
Additionally, do we need to add the args if we've already populated the command with them?

Reply:
> It seems like adding `s.args` to the cache key would result in the same problem we have right now where all of the args are in the cache key all the time.

In this PR, the args are not added to the cache key. They are used to replace args in the command. The key is constructed using only the evaluated command [here](https://github.com/GoogleContainerTools/kaniko/blob/ae5a7bf1586d10e95e4f7a5df56148bb8e079908/pkg/executor/build.go#L160)

Reply:
> I'm confused about the difference between `StageBuilider.args` and `StageBuilder.opts.args`

`StageBuilder.args` are union of  `KanikoOptions.args` i.e. `StageBuilder.opts.args` and any arg variables added in that stage with `ARG` command.
https://github.com/GoogleContainerTools/kaniko/blob/ae5a7bf1586d10e95e4f7a5df56148bb8e079908/pkg/executor/build.go#L132

Reply:
> Additionally, do we need to add the args if we've already populated the command with them?

We don't add the key. The line you are looking at is a different func `optimize` and not part of `populateCompositeKey`

Reply:
> > It seems like adding `s.args` to the cache key would result in the same problem we have right now where all of the args are in the cache key all the time.
> 
> In this PR, the args are not added to the cache key. They are used to replace args in the command. The key is constructed using only the evaluated command [here](https://github.com/GoogleContainerTools/kaniko/blob/ae5a7bf1586d10e95e4f7a5df56148bb8e079908/pkg/executor/build.go#L160)

Ahh ok, I see why I was confused now. Might be more clear to move the command resolve out of `populateCompositeKey`, but might be easier for maintenance to keep it in there.